src/extension/infer.rs

```diff
@@ -663,31 +676,98 @@ impl UnificationContext {
+                        match c {
+                            Constraint::Plus(_, other) => {
+                                newer_variables.insert(self.resolve(*other))
+                            }
+                            Constraint::Equal(other) => {
+                                newer_variables.insert(self.resolve(*other))
+                            }
+                        };
+                    }
+                    seen.insert(m);
+                }
+            }
+            new_variables = newer_variables;
+        }
+        seen
+    }
+
     /// Instantiate all variables in the graph with the empty extension set, or
     /// the smallest solution possible given their constraints.
     /// This is done to solve metas which depend on variables, which allows
     /// us to come up with a fully concrete solution to pass into validation.
+    ///
+    /// Nodes which loop into themselves must be considered as a "minimum" set
+    /// of requirements. If we have
+    ///   1 = 2 + X, ...
+    ///   2 = 1 + x, ...
+    /// then 1 and 2 both definitely contain X, even if we don't know what else.
+    /// So instead of instantiating to the empty set, we'll instantiate to `{X}`
     pub fn instantiate_variables(&mut self) {
-        for m in self.variables.clone().into_iter() {
+        // A directed graph to keep track of `Plus` constraint relationships
+        let mut relations = GraphContainer::<Directed>::new();
+        let mut solutions: HashMap<Meta, ExtensionSet> = HashMap::new();
+
+        let variable_scope = self.search_variable_deps();
+        // If `m` has been merged, [`self.variables`] entry
+        // will have already been updated to the merged
+        // value by [`self.merge_equal_metas`] so we don't
+        // need to worry about resolving it.
+        for m in variable_scope.into_iter() {
             if !self.solved.contains_key(&m) {
                 // Handle the case where the constraints for `m` contain a self
                 // reference, i.e. "m = Plus(E, m)", in which case the variable
                 // should be instantiated to E rather than the empty set.
-                let solution =
-                    ExtensionSet::from_iter(self.get_constraints(&m).unwrap().iter().filter_map(
-                        |c| match c {
-                            // If `m` has been merged, [`self.variables`] entry
-                            // will have already been updated to the merged
-                            // value by [`self.merge_equal_metas`] so we don't
-                            // need to worry about resolving it.
-                            Constraint::Plus(x, other_m) if m == self.resolve(*other_m) => {
-                                Some(x.clone())
-                            }
+                let plus_constraints =
+                    self.get_constraints(&m)
+                        .unwrap()
+                        .iter()
+                        .cloned()
```

Comment:
You probably don't need `.cloned()`? I'd expect you can do something on the next line like `|c| match *c {` or maybe just a `*other_m` on the line after

Reply:
We need to copy the meta and clone the `r`, so it amounts to the same thing?

Reply:
Ah yes, didn't see the `r`, fair point